### PR TITLE
Hashtag scraping

### PR DESCRIPTION
**Pull request by v2thegreat**

### CODE REVIEW

File: hashtag_scraper/tweepy_streamer.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+from tweepy.streaming import StreamListener
+from tweepy import OAuthHandler
+from tweepy import Stream
+
+
+from hashtag_scraper import twitter_credentials
+
+
+
+class TwitterClient():
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:11]: Add doc string for class

File: hashtag_scraper/tweepy_streamer.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    def __init__(self, twitter_user=None):
+        self.auth = TwitterAuthenticator().authenticate_twitter_app()
+        self.twitter_client = API(self.auth)
+        self.twitter_user = twitter_user
+
+    def get_twitter_client_api(self):
+        return self.twitter_client
+
+
+class TwitterAuthenticator():
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:22]: Add doc string for class

File: hashtag_scraper/tweepy_streamer.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    def authenticate_twitter_app(self):
+        auth = OAuthHandler(twitter_credentials.CONSUMER_KEY, twitter_credentials.CONSUMER_SECRET)
+        auth.set_access_token(twitter_credentials.ACCESS_TOKEN, twitter_credentials.ACCESS_TOKEN_SECRET)
+        return auth
+
+
+class TwitterStreamer():
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:30]: Add doc string for class

File: hashtag_scraper/tweepy_streamer.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        # This handles Twitter authentication and the connection to the Twitter Streaming API.
+        listener = TwitterListener()
+        auth = self.twitter_authenticator.authenticate_twiiter_app()
+        stream = Stream(auth, listener)
+
+        # this filters tweets, inside track can be a list of stuff to include inside the listener
+        stream.filter(track=hash_tag_list)
+
+
+class TwitterListener(StreamListener):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:45]: Add doc string for class

File: hashtag_scraper/tweepy_streamer.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+
+class TwitterClient():
+
+    def __init__(self, twitter_user=None):
+        self.auth = TwitterAuthenticator().authenticate_twitter_app()
+        self.twitter_client = API(self.auth)
+        self.twitter_user = twitter_user
+
+    def get_twitter_client_api(self):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:18]: Add docstring for function

File: hashtag_scraper/tweepy_streamer.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        return self.twitter_client
+
+
+class TwitterAuthenticator():
+
+    def authenticate_twitter_app(self):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:24]: Add docstring for function

File: hashtag_scraper/tweepy_streamer.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        auth.set_access_token(twitter_credentials.ACCESS_TOKEN, twitter_credentials.ACCESS_TOKEN_SECRET)
+        return auth
+
+
+class TwitterStreamer():
+
+    def __init__(self):
+        self.twitter_authenticator = TwitterAuthenticator
+
+    def stream_tweets(self, fetched_tweet_filename, hash_tag_list):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:35]: Add docstring for function

File: hashtag_scraper/tweepy_streamer.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        # this filters tweets, inside track can be a list of stuff to include inside the listener
+        stream.filter(track=hash_tag_list)
+
+
+class TwitterListener(StreamListener):
+
+    def __init__(self, fetched_tweets_filename):
+        self.fetched_tweets_filename = fetched_tweets_filename
+
+    def on_data(self, data):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:50]: Add docstring for function

File: hashtag_scraper/tweepy_streamer.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        try:
+            print(data)
+            with open(self.fetched_tweets_filename, 'a') as tf:
+                tf.write(data)
+            return True
+        except BaseException as e:
+            print("Error on data: %s" % str(e))
+        return True
+
+    def on_error(self, status):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:60]: Add docstring for function

File: hashtag_scraper/tweepy_streamer.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+        # Debug => prints text content of tweet of specified index
+        # print(temp[0].text)
+
+        for j in range(0, num_tweets):
+            tweets.append(temp[j].text)
+        trending[i] = tweets
+
+    print(trending)
+
```
[the remark this answers is not in the document]
Comment (RamKPrasanna) [reply hashtag_scraper/tweepy_streamer.py]: done

File: hashtag_scraper/tweepy_streamer.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        Modular function to return authentication keys to API function in the tweepy library.
+
+        :return: starts the authentication process by calling TwitterAuthenticator class
+        """
+        return self.twitter_client
+
+
+"""
+This class contains a function to fetch authentication keys from 'twitter_credentials.py' and serve along side the TwitterClient class.
+"""
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:95]: Move this to the class according to these [guidelines](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) (search for the ExampleClass). 

Note: we're using PEP8, not the google style sheet. This is only to showcase the position that the docstring needs to be

File: hashtag_scraper/tweepy_streamer.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        :return: Fetches Twitter API keys from twitter_credentials and uses OAuthHandler from tweepy library to authenticate Twitter API.
+        """
+        auth = OAuthHandler(twitter_credentials.CONSUMER_KEY, twitter_credentials.CONSUMER_SECRET)
+        auth.set_access_token(twitter_credentials.ACCESS_TOKEN, twitter_credentials.ACCESS_TOKEN_SECRET)
+        return auth
+
+
+"""
+This class automatically passes the TwitterAuthenticator and TwitterListener to connect to the Twitter API.
+"""
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:113]: Move this to the class according to these [guidelines](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) (search for the ExampleClass). 

Note: we're using PEP8, not the google style sheet. This is only to showcase the position that the docstring needs to be

File: hashtag_scraper/tweepy_streamer.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        auth = self.twitter_authenticator.authenticate_twiiter_app()
+        stream = Stream(auth, listener)
+
+        # this filters tweets, inside track can be a list of stuff to include inside the listener
+        stream.filter(track=hash_tag_list)
+
+
+"""
+This class listens to the response from the Twitter API for any important codes or exceptions that are raised. 
+"""
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:133]: Move this to the class according to these [guidelines](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) (search for the ExampleClass). 

Note: we're using PEP8, not the google style sheet. This is only to showcase the position that the docstring needs to be

File: hashtag_scraper/tweepy_streamer.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
+from tweepy import API
+from tweepy.streaming import StreamListener
+from tweepy import OAuthHandler
+from tweepy import Stream
+import re
+
+from hashtag_scraper import twitter_credentials
+
+
+def clean_tweet(tweet):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:10]: convert to private functions (rename to _clean_tweet)

Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:10]: Add type hinting for the parameter `tweet`

File: hashtag_scraper/tweepy_streamer.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    """
+    Cleans the tweet using regular expression for special characters or hyperlinks.
+
+    :param tweet: text to be processed
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:14]: Add type hinting for the parameter `tweet`

File: hashtag_scraper/tweepy_streamer.py [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    :return: text without special characters or hyperlinks
+    """
+    return ' '.join(re.sub("(@[A-Za-z0-9]+)|([^0-9A-Za-z\t])|(\w+:\/\/\S+)", " ", tweet).split())
+
+
+def fetch_tweets(num_topics, num_tweets, geo_code=2388929, clean_bool=True):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:20]: Add type hinting for the parameter `num_topics`, `num_tweets`, `geo_code`, `clean_bool`

File: hashtag_scraper/tweepy_streamer.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    """
+    Fetches a number of tweets on a specified number of trending topics using a particular geo code.
+
+    :param num_topics: Number of treading topics to be collected.
+    :param num_tweets: Number of tweets per topic to be collected.
+    :param geo_code: Geo code to get trending topics for that area.
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:26]: Add type hinting for the parameter `num_topics`, `num_tweets`, `geo_code`, `clean_bool`

File: hashtag_scraper/tweepy_streamer.py (hunk 16 above)
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:20]: rename `clean_bool` to `clean_tweets` for better understanding

File: hashtag_scraper/tweepy_streamer.py [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+
+
+"""
+This class contains basic functions to authenticate using the TwitterAuthenticator class. 
+"""
+
+
+class TwitterClient():
+
+    def __init__(self, twitter_user=None):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:79]: Add type hinting for the parameters and docstring

File: hashtag_scraper/tweepy_streamer.py [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+
+
+"""
+This class listens to the response from the Twitter API for any important codes or exceptions that are raised. 
+"""
+
+
+class TwitterListener(StreamListener):
+
+    def __init__(self, fetched_tweets_filename):
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:138]: Add docstring for function and type hinting

File: hashtag_scraper/tweepy_streamer.py [hunk 20]
```diff
@@ -0,0 +1,169 @@
+from tweepy import API
+from tweepy.streaming import StreamListener
+from tweepy import OAuthHandler
+from tweepy import Stream
+import re
+
+from hashtag_scraper import twitter_credentials
```
Comment (v2thegreat) [inline hashtag_scraper/tweepy_streamer.py:7]: File `twitter_credentials` does not exist

**Pull request merged**